File: Java-Project/DataStructureNonLinear/src/datastructure/graph_undirected/WeightedGraph.java
package datastructure.graph_undirected;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;
import java.util.Stack;

public class WeightedGraph { // Weighted
	private class Node {
		public String lable;
		public List<Edge> edges; // better Map
		public Node(String lable) { this.lable = lable; edges = new ArrayList<>(); }
		public void addEdge(Node to, int weight) { edges.add(new Edge(this, to, weight)); }
		@Override
		public String toString() { return lable; }
	}
	private class NodePriority {
		public Node node;
		public int priority;
		public NodePriority(Node node, int priority) { this.node = node; this.priority = priority; }
	}

	private class Edge {
		public Node from;
		public Node to;
		public int weight;
		public Edge(Node from, Node to, int weight) { this.from = from; this.to = to; this.weight = weight; }
		@Override
		public String toString() { return from + " " + weight + "> " + to; } // A->B
	}

	private Map<String, Node> nodes;

	/// private Map<Node, List<Edge>> adjecencyList;
	public WeightedGraph() { nodes = new HashMap<>(); } // adjecencyList = new HashMap<>();

	private boolean isNull(Node node) { return node == null; }

	private boolean isNull(Edge edge) { return edge == null; }

	public void addNode(String lable) {
		nodes.putIfAbsent(lable, new Node(lable));
		// adjecencyList.putIfAbsent(node, new ArrayList<>());
		// Edges will initialize automatically
	}
	/*
	 * public void removeNode(String label) { var node = nodes.get(label);
	 * if(isNull(node)) return; for(var n : adjecencyList.keySet())
	 * adjecencyList.get(n).remove(node); adjecencyList.remove(node);
	 * nodes.remove(node); }
	 */

	public void addEdge(String from, String to, int weight) { // relationship
		var fromNode = nodes.get(from);
		var toNode = nodes.get(to);
		if (isNull(fromNode) || isNull(toNode))
			throw new IllegalArgumentException();
		fromNode.addEdge(toNode, weight);
		toNode.addEdge(fromNode, weight);
		// adjecencyList.get(fromNode).add(new Edge(fromNode, toNode, weight));
		/// *!*/adjecencyList.get(toNode).add(new Edge(toNode, fromNode, weight));
	}

	public Path getShortestPath(String from, String to) {
		var fromNode = nodes.get(from);
		var toNode = nodes.get(to);
		if (isNull(fromNode)) throw new IllegalArgumentException();
	    if (isNull(toNode)) throw new IllegalArgumentException();
		Map<Node, Integer> distances = new HashMap<>();
		for (var node : nodes.values()) distances.put(node, Integer.MAX_VALUE);
		Map<Node, Node> previousNodes = new HashMap<>();
		Set<Node> visited = new HashSet<>();
	    PriorityQueue<NodePriority> queue = new PriorityQueue<>( Comparator.comparingInt(ne -> ne.priority));
	    			//queue, PriorityQueue with  Comparator
	    queue.add(new NodePriority(fromNode, 0));	// only item in the queue

		distances.replace(fromNode, 0); // A 0
	    while (!queue.isEmpty()) { 
	    	var current = queue.remove().node;
	    	visited.add(current);

	    	for (var edge : current.edges) {
	    		if (visited.contains(edge.to)) continue;
	    		var newDistance = distances.get(current) + edge.weight;
		        if (newDistance < distances.get(edge.to)) {
		          distances.replace(edge.to, newDistance);
		          previousNodes.put(edge.to, current);
		          queue.add(new NodePriority(edge.to, newDistance));
		        }
	    	}
	    }
	    //return distances.get(toNode);
	    return buildPath(previousNodes,toNode);
	}
	private Path buildPath(Map<Node, Node> previousNodes, Node toNode){
		    Stack<Node> stack = new Stack<>();
		    stack.push(toNode);
		    var previous = previousNodes.get(toNode);
		    while(!isNull(previous)) {
		    	stack.push(previous);
		    	previous = previousNodes.get(previous);
		    }
		    var path = new Path();
		    while(!stack.isEmpty()) path.add(stack.pop().lable);
		    return path;
	}
	@Override
	public String toString() {
		String str = "";
		for (var node : nodes.values()/* adjecencyList.keySet() */) {
			var targets = node.edges;// adjecencyList.get(source);
			if (!targets.isEmpty())
				str += node + " is connected to " + targets + "\n";
		}
		return str;
	}
}

File: CSharp-Project/DataStructure/GraphUndirected/WeightedGraph.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GraphUndirected
{
    public class WeightedGraph
    { // Weighted
        private class Vertice   //Node
        {
            public String Lable { get; private set; }
     /*!*/  public List<Edge> EdgeList { get; private set; } // better Map
            public Vertice(String lable) { Lable = lable; EdgeList = new(); }
     /*!*/  public void AddEdge(Vertice to, int weight) { EdgeList.Add(new Edge(/*this,*/ to, weight)); } //
            public override String ToString() { return Lable; }
        }

        private class Edge
        {
           // public Vertice From { get; private set; }
            public Vertice To { get; private set; }
            public int Weight { get; private set; }
            public Edge(/*Vertice from,*/ Vertice to, int weight) { /*From = from;*/ To = to; Weight = weight; }
            public override String ToString() { return /*From +*/ "<" + Weight + ">" + To; } // A->B
        }
/*        private class NodePriority
        {
            public Vertice Node { get; set; }
            public int Priority { get; set; }
            public NodePriority(Vertice node, int priority) { Node = node; Priority = priority; }
        }
*/

        private Dictionary<String, Vertice> nodes;

        /// private Map<Node, List<Edge>> adjecencyList;
        public WeightedGraph() { nodes = new(); }

        private bool IsNull(Vertice node) { return node == null; }

        public void AddNode(String lable) { nodes.TryAdd(lable, new Vertice(lable)); }

        public void AddEdge(String fromString, String toString, int weight)
        { // relationship
            var from = nodes?[fromString];
            var to = nodes?[toString];
            if (IsNull(from!) || IsNull(to!)) throw new Exception();
            from!.AddEdge(to!, weight);
            to!.AddEdge(from!, weight);
        }



        public List<String> GetShortestPath(String fromString, String toString)
        {
            var fromNode = nodes?[fromString];
            var toNode = nodes?[toString];
            if (IsNull(fromNode!) || IsNull(toNode!)) throw new Exception();

            Dictionary<Vertice, int> distances = new();
            foreach (var node in nodes!.Values) distances.Add(node, int.MaxValue);
            Dictionary<Vertice, Vertice> previousNodes = new();
            HashSet<Vertice> visited = new();
            PriorityQueue<Vertice, int> queue = new ();
           //java: PriorityQueue<NodeEntry> queue = new PriorityQueue<>(Comparator.comparingInt(ne->ne.priority));
            queue.Enqueue(fromNode!, 0);  // only item in the queue

            distances[fromNode] = 0; // java: replace(,) // A 0
            while (queue.Count != 0)
            {
                var current = queue.Dequeue();
                visited.Add(current);

                foreach (var edge in current.EdgeList)
                {
                    if (visited.Contains(edge.To)) continue;
                    var newDistance = distances[current] + edge.Weight;
                    if (newDistance < distances[edge.To])
                    {
                        distances[edge.To] = newDistance; // java: replace(,)
                        previousNodes.Add(edge.To, current);
                        queue.Enqueue(edge.To, newDistance);
                    }
                }
            }
            //return distances.get(toNode);
            return buildPath(previousNodes, toNode!);
        }
        private List<String> buildPath(Dictionary<Vertice, Vertice> previousNodes, Vertice toNode)
        {
            Stack<Vertice> stack = new();
            stack.Push(toNode);
            var previous = previousNodes?[toNode];
            while (true) //!IsNull(previous!)
            {
                stack.Push(previous!);
                try { previous = previousNodes![previous!]; }
                catch (Exception) { break; } //previous = previousNodes.GetValueOrDefault(previous, null);
            }
            return ToList(stack);
        }

        private List<String> ToList(Stack<Vertice> stack)
        {
            List<String> NodeList = new();
            foreach (var node in stack) NodeList.Add(node.Lable);
            return NodeList;
        }

        public override String ToString()
        {
            StringBuilder str = new("");
            foreach (var node in nodes.Values/* adjecencyList.keySet() */)
            {
                var targets = node.EdgeList;// adjecencyList.get(source);
                if (targets.Count != 0) str.Append(node + " is connected to [" + String.Join(", ", targets) + "]\n");
            }
            return str.ToString();
        }

    }
}
